### DISCUSSION TITLE
determining time of fixation of a fixed mutation

### DISCUSSION THREAD

**Discussion by mawassw** [tskit-dev/tskit#2664] [Q&A]
Hi everyone,

I am wondering if it is possible to determine the time of fixation of a fixed mutation. Here are some details to give context for my analysis. I am using the tskit C API in our forward-time evolutionary simulator, so I can easily obtain the effect size of a mutation and the time of appearance of a mutation through the mutation table. However, for the purpose of my project I need to figure out the time of fixation of a fixed mutation. Is this possible using the information stored in the tree sequences? 

P.S. to figure out if a mutation fixed for a population that is in quasi-equilibirum, I extract using the Python API the mutations that appeared at or before the single root of a tree at a given site. I am not sure if this is a right approach however, so I would appreciate input on this issue as well.

TLDR: How can I extract the fixed mutations and their time of fixation using tskit.

I can provide more information and details if necessary
Thank you!

**Comment by petrelharp** [comment]
Hello, Walid! Let's see - this all depends on what you are recording in the tree sequence with your simulator. If you remember everyone, ever, then you can find out everything you want. But, if you are running `simplify` (as is usual so the tree sequence does not get huge), then you won't know the time of fixation - the only things that you can know is (a) the time of appearance of the mutation (if you record this somehow, eg in metadata), and (b) the time of the MRCA of everyone alive today that carried the mutation. This is because, fundamentally, the tree sequence is designed to record the history of a set of samples - probably in your case the genomes alive today - and many of the genomes at other points back in time are not needed to record that history.

**Reply by jeromekelleher** [reply]
> (a) the time of appearance of the mutation (if you record this somehow, eg in metadata),

Just a quick note here that there is a ``time`` column in the mutation table for this.

**Reply by petrelharp** [reply]
Whoops, right, heh. =)

**Reply by mawassw** [reply]
Hi Peter & Jerome,

in terms of (a) yes, I am using the time in the mutation table to get the time of appearance of a mutation, and I know how to get (b). I appreciate your answer Peter because it confirms my initial suspicion that I have to avoid simplifying if I want to get the time of fixation of a mutation that fixed during the simulation. 

If we presume that I managed to run my simulation without simplification, how would I go about finding the fixation time? If I understand correctly from your answer using the time of the MRCA of everyone alive that carried the mutation does not amount for the time of fixation. But would the mutation that is carried by the MRCA be considered a mutation that is fixed in the final sample?

Thank you so much for both your comments!

**Reply by jeromekelleher** [reply]
I'm not as well versed as Peter in forward sims, but one (probably bad) thing you could do is run simplification every generation. Then the first time you see a mutation getting pushed over the (single) root of its tree is its fixation time.

**Reply by petrelharp** [reply]
I think you just basically have to compute allele frequencies every generation (well, just check if they are equal to 100%). Simplifying every generation and seeing if the mutation is over the root is one way to do that - it would be kinda expensive, but at least would mean you don't have to keep around everyone-ever? If you have the full everyone-ever tree sequence, I don't know a clever way off the top of my head to figure this out, but you could use the snippet of code to compute allele frequencies with sample sets equal to <whoever is alive at time t> (see the github issue, sorry, I don't know where it is right now).

**Reply by hyanwong** [reply]
If you have a fixed population size, you could simply look to see when the number of lineages first hits that (haploid) size.

**Reply by mawassw** [reply]
Thanks everyone for your responses! 

@hyanwong Unfortunately, I am not working with a fixed population size because we are testing our evolutionary model with an absolute fitness scheme. Other people in my group are using a relative fitness scheme so they have a fixed population and they use the exact method you mentioned. Thank you!

@petrelharp @jeromekelleher  Thank you for that suggestion! I am testing out right now if it is possible to run without simplification since we have access to a high memory HPC. MY group and I came up with this solution: at origination, there are N-1 lineages that lack the mutation, so we run a script for a given mutation, the max time to get to N-1 lineages that do have it. Any thoughts on this approach?
if the plan to not simplify is not successful then most likely we will opt to use your suggestion of simplifying at every generation and check until a mutation is over the single root of a tree.
